### PR TITLE
v0.3.0

### PR DESCRIPTION
**Pull request by xlwknx**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,13 @@
 # Changelog
 
+## 2018-11-30 0.3.0
+
+* Removed method EThree.boostrap(pwd)
+* Method EThree.bootstrap() renamed to EThree.register()
+* New method EThree.rotatePrivateKey() - used in case if you lost your private key
+* New method EThree.restorePrivateKey(pwd) - used to fetch private key from Virgil Cloud
+* New method EThree.hasLocalPrivateKey() - used to check private key existance on a device
```
Comment (vadimavdeev) [inline CHANGELOG.md]:
```suggestion
* New method EThree.hasLocalPrivateKey() - used to check for private key existence on a device
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,13 @@
 # Changelog
 
+## 2018-11-30 0.3.0
+
+* Removed method EThree.boostrap(pwd)
+* Method EThree.bootstrap() renamed to EThree.register()
+* New method EThree.rotatePrivateKey() - used in case if you lost your private key
```
Comment (vadimavdeev) [inline CHANGELOG.md]:
```suggestion
* New method EThree.rotatePrivateKey() - used in case you lost your private key
```

File: src/EThree.ts [hunk 3]
```diff
@@ -1,58 +1,100 @@
...
-        if (hasCard) {
-            if (!password) {
-                throw new PasswordRequiredError();
-            } else {
-                await this.keyLoader.loadRemotePrivateKey(password);
-                return;
-            }
-        } else {
+    async register() {
+        if (this.inProcess) throwMultithreadError(this.register.name);
```
Comment (vadimavdeev) [inline src/EThree.ts]: Лучше имя просто строкой передать. Свойство `name` у функций не поддерживается например в IE

File: src/EThree.ts [hunk 4]
```diff
@@ -1,58 +1,100 @@
...
 
 type EncryptVirgilPublicKeyArg = VirgilPublicKey[] | VirgilPublicKey;
 
+interface IEThreeOptions {
+    provider: CachingJwtProvider;
+    toolbox?: VirgilToolbox;
+    keyLoader?: PrivateKeyLoader;
+}
+
+const throwMultithreadError = (method: string) => {
```
Comment (vadimavdeev) [inline src/EThree.ts]: Even though this function is local to the module, the name is misleading. It is not a threading issue that this method is preventing. 
```suggestion
const throwIllegalInvocationError = (method: string) => {
```

File: src/EThree.ts [hunk 5]
```diff
@@ -1,58 +1,100 @@
...
+        if (this.inProcess) throwMultithreadError(this.rotatePrivateKey.name);
+        this.inProcess = true;
+        try {
+            const [cards, privateKey] = await Promise.all([
+                this.toolbox.cardManager.searchCards(this.identity),
+                this.keyLoader.loadLocalPrivateKey(),
+            ]);
+            if (cards.length === 0) throw new RegisterRequiredError();
+            if (cards.length > 1) throw new MultipleCardsError(this.identity);
+            if (privateKey) throw new PrivateKeyAlreadyExistsError();
```
Comment (vadimavdeev) [inline src/EThree.ts]: Shouldn't this method just overwrite the local private key if it exists?

File: src/EThree.ts [hunk 6]
```diff
@@ -1,58 +1,100 @@
...
+            await this.toolbox.publishCard(keyPair, cards[0].id);
+            await this.keyLoader.savePrivateKeyLocal(keyPair.privateKey);
+        } finally {
+            this.inProcess = false;
+        }
+    }
+
+    async restorePrivateKey(pwd: string): Promise<void> {
+        try {
+            await this.keyLoader.restorePrivateKey(pwd);
```
Comment (vadimavdeev) [inline src/EThree.ts]: why not checking `hasLocalPrivateKey()` before trying to restore?

File: src/PrivateKeyLoader.ts [hunk 7]
```diff
@@ -1,60 +1,37 @@
...
+    async savePrivateKeyRemote(privateKey: VirgilPrivateKey, password: string) {
+        const storage = await this.getStorage(password);
         return await storage.storeEntry(
             this.identity,
             this.toolbox.virgilCrypto.exportPrivateKey(privateKey),
         );
     }
 
-    async savePrivateKeyLocal(privateKey: VirgilPrivateKey) {
+    async savePrivateKeyLocal(privateKey: VirgilPrivateKey, isPublished?: boolean) {
```
Comment (vadimavdeev) [inline src/PrivateKeyLoader.ts]: what do you need `isPublished` for?

File: src/PrivateKeyLoader.ts [hunk 8]
```diff
@@ -68,80 +45,55 @@ export default class PrivateKeyLoader {
...
             if (typeof e === 'object' && e.code === 60007) {
                 const promise = new Promise((resolve, reject) => {
                     const repeat = () =>
-                        this.brainKey
-                            .generateKeyPair(password)
+                        brainKey
+                            .generateKeyPair(pwd)
                             .then(resolve)
                             .catch(reject);
                     setTimeout(repeat, 2000);
```
Comment (vadimavdeev) [inline src/PrivateKeyLoader.ts]: `2000` should be a constant `BRAIN_KEY_RATE_LIMIT_DELAY` or something like that

File: src/PrivateKeyLoader.ts [hunk 9]
```diff
@@ -150,4 +102,25 @@ export default class PrivateKeyLoader {
...
+                this.toolbox.jwtProvider,
+                keyPair.privateKey,
+                keyPair.publicKey,
+                undefined,
+                new KeyknoxCrypto(this.toolbox.virgilCrypto),
+            ),
+        );
+        try {
+            await storage.retrieveCloudEntries();
+        } catch (e) {
```
Comment (vadimavdeev) [inline src/PrivateKeyLoader.ts]: add a check to see if the error occurred because of the wrong password

File: src/errors.ts [hunk 10]
```diff
@@ -39,25 +36,76 @@ export class EmptyArrayError extends SdkError {
     }
 }
 
-const isError = (arg: any): arg is Error => arg instanceof Error;
-const isPublicKey = (arg: any): arg is VirgilPublicKey => !(arg instanceof Error);
+export class PrivateKeyAlreadyExistsError extends SdkError {
+    constructor() {
+        super(
+            'You already have a private key. Use EThree.cleanup() to delete it. If you delete last copy of private key, you will not able to decrypt any information encrypted for this private key',
```
Comment (vadimavdeev) [inline src/errors.ts]:
```suggestion
            'You already have a private key. Use EThree.cleanup() to delete it. If you delete the last copy of the private key, you will not be able to decrypt any information encrypted for this private key',
```

File: src/errors.ts [hunk 11]
```diff
@@ -39,25 +36,76 @@ export class EmptyArrayError extends SdkError {
...
+            'PrivateKeyAlreadyExistsError',
+        );
+    }
+}
+
+export class PrivateKeyNoBackupError extends SdkError {
+    constructor() {
+        super("Backup private key doesn't exist", 'PrivateKeyNoBackupError');
```
Comment (vadimavdeev) [inline src/errors.ts]:
```suggestion
        super("Backup copy of private key doesn't exist", 'PrivateKeyNoBackupError');
```

File: src/errors.ts [hunk 12]
```diff
@@ -39,25 +36,76 @@ export class EmptyArrayError extends SdkError {
...
+    }
+}
+
+export class MultipleCardsError extends SdkError {
+    constructor(identity: string) {
+        super(
+            `There are several public keys registered with ${identity}, which is not supported.`,
```
Comment (vadimavdeev) [inline src/errors.ts]: What to do in this case? Choose a different identity?

File: src/errors.ts [hunk 13]
```diff
@@ -39,25 +36,76 @@ export class EmptyArrayError extends SdkError {
...
+            }
+        }
+        return result;
     }
 
-    constructor(result: Array<VirgilPublicKey | Error>) {
+    constructor(identities: string[], result: Array<VirgilPublicKey | Error>) {
         super(
-            'Some promises got rejected. Use err.rejected for unhandled results, .resolved for handled and .result for all responses',
+            `Failed some public keys lookups. You can see the results by error.resolved() and error.rejected() methods of this error instance`,
```
Comment (vadimavdeev) [inline src/errors.ts]:
```suggestion
            `Failed some public keys lookups. You can see the results by calling error.resolved() and error.rejected() methods of this error instance`,
```

File: src/errors.ts (hunk 12 above)
Comment (xlwknx) [reply src/errors.ts]: For now yes, but I'm think maybe show the latest card?

Comment (xlwknx) [reply src/errors.ts]: Yea, we decide to throw error in all sdk, cause it is unexpected situation

File: src/PrivateKeyLoader.ts [hunk 14]
```diff
@@ -68,86 +49,83 @@ export default class PrivateKeyLoader {
...
+            if (e instanceof VirgilCryptoError) throw new WrongKeyknoxPasswordError();
```
Comment (vadimavdeev) [inline src/PrivateKeyLoader.ts]: please check that `e.name === 'VirgilCryptoError'` instead. Currently `VirgilCryptoError` is an internal type.

**Pull request merged**